File: HW_2/ship.cpp
#include <cstdlib>
#include <cstring>
#include "ship.h"


int RandInt(int min, int max);

char *ReadString(FILE *file);

const char *liner_name = "liner";
const char *tow_name = "tow";
const char *tanker_name = "tanker";

// Read ship information from file.
void ship::Read(FILE *file) {
    transport::Read(file);
    fscanf(file, "%d", &this->displacement);
    char *current_type = ReadString(file);

    if (!strcmp(current_type, liner_name)) {
        this->type = shipType::LINER;
    } else if (!strcmp(current_type, tow_name)) {
        this->type = shipType::TOW;
    } else if (!strcmp(current_type, tanker_name)) {
        this->type = shipType::TANKER;
    } else {
        exit(1);
    }
    free(current_type);
}

// Write ship information to file.
void ship::Write(FILE *file) {
    fprintf(file, "%s", "ship ");
    transport::Write(file);
    fprintf(file, "%c%d%c", ' ', this->displacement, ' ');
    if (this->type == shipType::TANKER) {
        fprintf(file, "%s", tanker_name);
    } else if (this->type == shipType::TOW) {
        fprintf(file, "%s", tow_name);
    } else {
        fprintf(file, liner_name);
    }
}

// Generate random ship.
void ship::Generate() {
    transport::Generate();
    this->displacement = RandInt(10, 350);
    this->type = (shipType)(rand() % 3);
}

File: HW_3/Ship.py
from Transport import Transport
import enum
import random


class ShipType(enum.Enum):
    LINER = 1
    TOW = 2
    TANKER = 3


str_to_type = {"LINER": ShipType.LINER, "TOW": ShipType.TOW, "TANKER": ShipType.TANKER}


class Ship(Transport):
    def __init__(self, params=None):
        super().__init__(params)
        if params is None:
            self.type = ShipType.LINER
            self.displacement = 0
        else:
            self.displacement = int(params[2])
            self.type = str_to_type[params[3]]

    def generate(self):
        super().generate()
        self.type = ShipType(random.randint(1, 3))
        self.displacement = random.randint(100, 1000)
        return self

    def __str__(self):
        return f"Ship {super().__str__()} {self.displacement} {self.type.name}"
